### PR TITLE
Fix/batch norm

### PR DESCRIPTION
**Pull request by nathanielsimard**
Fix multiple problems that added up to bugs with batch norm:

- Fix batch norm 2d (problem with batch size reshaping) modified the tests with batch_size > 1 (that failed with the previous implementation).
- The `tch` crate is unsafe to use, even if not explicitly said so. Therefore, instead of wrapping the tensor handle inside an `Arc` and using the Rust rules for internal mutability, I'm wrapping the data pointer. I added some tests (that failed before the changes) where the original tensor would be changed even if multiple references were pointing to it.
- Updated the MNIST example to use batch norm.

### CODE REVIEW

File: burn-tch/Cargo.toml [hunk 1]
```diff
@@ -15,6 +15,8 @@ doc = ["tch/doc-only"]
 
 [dependencies]
 burn-tensor = {path = "../burn-tensor"}
+libc = "0.2.0"
```
Comment (antimora) [inline burn-tch/Cargo.toml:18]: Going sublevel, aren't we =)

File: examples/mnist/examples/mnist.rs [hunk 2]
```diff
@@ -24,9 +24,15 @@ mod tch_gpu {
     pub fn run() {
         #[cfg(not(target_os = "macos"))]
         let device = TchDevice::Cuda(0);
+        #[cfg(not(target_os = "macos"))]
+        type FloatElement = burn::tensor::f16;
```
Comment (antimora) [inline examples/mnist/examples/mnist.rs:28]: We need to test f16 on GPU without f16 native support. My linux  GPU (GeForce GTX 1080) does not have f16 native.  Maybe it works without overflow.

**Review by antimora — CHANGES_REQUESTED** [review]
Wow, so many changes to fix one bug =)

Overall, it's looking good. I found a few minor issues that I could identify.

File: examples/mnist/examples/mnist.rs (hunk 2 above)
Comment (nathanielsimard) [reply examples/mnist/examples/mnist.rs:28]: I'll put f32 by default then, it's not a big deal :)

**Review by antimora — APPROVED** [review]
Looks good. 

I am running some training to synchronize models on for the web . It looks the number of params got doubled (to 633407). If it's okay with you, I'll adjust hyperparams a little.

File: burn-tch/src/tensor.rs [hunk 3]
```diff
@@ -38,39 +69,51 @@ unsafe impl<E: tch::kind::Element, const D: usize> Send for TchTensor<E, D> {}
...
+    pub fn mut_ops<
+        F: Fn(&mut tch::Tensor) -> tch::Tensor,
+        EOut: tch::kind::Element,
+        const D_OUT: usize,
+    >(
+        &mut self,
+        func: F,
+    ) -> Option<TchTensor<EOut, D_OUT>> {
+        if Arc::strong_count(&self.storage) > 1 {
+            return None;
```
Comment (twitchax) [inline burn-tch/src/tensor.rs:82]: Nice.

**Review by twitchax — APPROVED** [review]
LGTM.

The tests are going to help ensure the mutability rules are conserved going forward, which is a good peace of mind.
